src/coprocessor/codec/batch/buffer_vec.rs

```diff
@@ -0,0 +1,800 @@
+                    std::ptr::copy(
+                        self.data.as_ptr().add(self.offsets[len - 1]),
+                        self.data.as_mut_ptr().add(data_write_offset),
+                        write_len,
+                    );
+                    self.offsets[offsets_write_offset] = data_write_offset;
+                    offsets_write_offset += 1;
+                    data_write_offset += write_len;
+                }
+                self.offsets.set_len(offsets_write_offset);
+                self.data.set_len(data_write_offset);
+            }
+        }
+    }
+
+    /// Returns an iterator over the `BufferVec`.
+    pub fn iter(&self) -> Iter<'_> {
+        Iter {
+            data: &self.data,
+            offsets: &self.offsets,
+        }
+    }
+}
+
+impl std::ops::Index<usize> for BufferVec {
+    type Output = [u8];
+
+    #[inline]
+    fn index(&self, index: usize) -> &[u8] {
+        assert!(index < self.offsets.len());
+        if index < self.offsets.len() - 1 {
+            &self.data[self.offsets[index]..self.offsets[index + 1]]
+        } else {
+            &self.data[self.offsets[index]..]
+        }
+    }
+}
+
+// TODO: Implement Index<XxxRange>
+
+#[derive(Clone, Copy)]
+pub struct Iter<'a> {
+    data: &'a [u8],
+    offsets: &'a [usize],
+}
+
+impl<'a> Iterator for Iter<'a> {
+    type Item = &'a [u8];
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.nth(0)
+    }
+
+    #[inline]
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let len = self.offsets.len();
+        (len, Some(len))
+    }
+
+    #[inline]
+    fn count(self) -> usize {
+        self.offsets.len()
+    }
+
+    fn last(mut self) -> Option<Self::Item> {
+        let l = self.offsets.len();
+        if l == 0 {
+            None
+        } else {
+            self.nth(l - 1)
+        }
+    }
+
+    fn nth(&mut self, n: usize) -> Option<Self::Item> {
+        if n + 1 < self.offsets.len() {
+            let begin_offset = self.offsets[n];
+            let end_offset = self.offsets[n + 1];
+            self.offsets = &self.offsets[n + 1..];
+            Some(&self.data[begin_offset..end_offset])
+        } else if n + 1 == self.offsets.len() {
+            let begin_offset = self.offsets[n];
+            self.offsets = &[];
+            Some(&self.data[begin_offset..])
+        } else {
+            self.offsets = &[];
+            None
+        }
+    }
+}
+
+impl<'a> ExactSizeIterator for Iter<'a> {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_basic() {
+        let mut v = BufferVec::new();
+        assert_eq!(format!("{:?}", v), "[]");
+        assert!(v.is_empty());
+
+        v.pop();
+        assert_eq!(format!("{:?}", v), "[]");
+        assert!(v.is_empty());
+
+        v.shift(0);
+        assert_eq!(format!("{:?}", v), "[]");
+        assert!(v.is_empty());
+
+        v.push(&[0xAA, 0x0, 0xB]);
+        assert_eq!(v.len(), 1);
```

Comment:
Can we abstract a function to do these assertions?

Reply:
The abstracted function may be more complicated than the thing it is testing.